src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -1087,14 +1109,21 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
    * typically happen after this has run, and we loop around to retry the alloc
    * if the state says we should.
    */
-  bool post_alloc_failed(long thread_id, bool is_oom) {
+  bool post_alloc_failed(long thread_id, bool is_oom, bool likely_spill) {
     std::unique_lock<std::mutex> lock(state_mutex);
     auto thread = threads.find(thread_id);
     // only retry if this was due to an out of memory exception.
     bool ret = true;
-    if (thread != threads.end()) {
+    if (!likely_spill && thread != threads.end()) {
       switch (thread->second.state) {
-        case TASK_ALLOC_FREE: transition(thread->second, thread_state::TASK_RUNNING); break;
+        case TASK_ALLOC_FREE: 
+          if (is_oom) {
+            transition(thread->second, thread_state::TASK_BLOCKED);
```

Comment:
Why are we blocking on an OOM if a free happened while in the middle of the allocation? That is the entire point of the TASK_ALLOC_FREE state.

Reply:
oh I am sorry, I think this was something I should have removed before pushing, since we have the `likely_spill` boolean that helps us skip this whole block in this recursive scenario.